Feature/fixes private notes

Rudeka opened the pull request:
Add userId to collaborators
Add note update if one already exists
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62653642%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62658153%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62658356%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62687273%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62688415%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62688613%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62691349%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62711876%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62713685%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62844614%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62877895%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62881279%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62881505%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62976586%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62976788%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r63161901%22%2C%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r63165588%22%5D%2C%20%22comments%22%3A%20%7B%22Pull%20d9c874074245403f0fae24e90d3b6e91f41e9c49%20DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/NoteCommands.cs%2019%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62881279%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22So%20it%20works%3F%20Did%20you%20verify%20it%3F%22%2C%20%22created_at%22%3A%20%222016-05-11T16%3A33%3A11Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6068068%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/TerraVenil%22%7D%7D%2C%20%7B%22body%22%3A%20%22Yes%2C%20it%20works.%20With%20EntityState%20I%20had%20to%20make%20it%20tracked%20first%20and%20then%20call%20method%20Remove.%22%2C%20%22created_at%22%3A%20%222016-05-12T07%3A43%3A18Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/14029898%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Rudeka%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/NoteCommands.cs%3AL19-36%22%7D%2C%20%22Pull%20b34981bfea3cfcb9c7e9f77f6b22c2c54ce4ed1d%20DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs%2031%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62688613%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Can%20you%20do%20this%20only%20by%20one%20call%20to%20DB%3F%22%2C%20%22created_at%22%3A%20%222016-05-10T15%3A02%3A36Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6068068%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/TerraVenil%22%7D%7D%2C%20%7B%22body%22%3A%20%22Fixed%20and%20commited%22%2C%20%22created_at%22%3A%20%222016-05-11T13%3A12%3A08Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/14029898%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Rudeka%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs%3AL209-235%22%7D%2C%20%22Pull%20d9c874074245403f0fae24e90d3b6e91f41e9c49%20DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs%2029%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62881505%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22To%20many%20works%20inside%20one%20method....Please%20refactor%20it.%22%2C%20%22created_at%22%3A%20%222016-05-11T16%3A34%3A29Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6068068%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/TerraVenil%22%7D%7D%2C%20%7B%22body%22%3A%20%22Moved%20query%20to%20context%20to%20action%20method%20and%20passed%20it%20as%20an%20argument.%20Commit%20added%20%22%2C%20%22created_at%22%3A%20%222016-05-12T07%3A41%3A43Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/14029898%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Rudeka%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs%3AL209-237%22%7D%2C%20%22Pull%2056d9f5d8a20945844b5c85d8ad461e6d4470cee9%20DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/INoteCommands.cs%205%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62653642%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Explain%20please%20why%20you%20are%20decided%20change%20method%20name%20AddNote%20-%3E%20AddOrUpdateNote%3F%20What%20about%20SOLID%3F%22%2C%20%22created_at%22%3A%20%222016-05-10T11%3A18%3A07Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6068068%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/TerraVenil%22%7D%7D%2C%20%7B%22body%22%3A%20%22We%20should%20have%20only%20one%20note%20for%20collaborator%20from%20user%20and%20I%20decided%20not%20to%20call%20separate%20method%20for%20update%20a%20note%2C%20if%20one%20already%20exists.%20It%20is%20not%20single%20responsibility%20and%20decided%20for%20this%20simple%20case%20we%20can%20ignore%20it.%5Cr%5CnBut%20now%20I%20think%20I%20should%20better%20have%20two%20methods%20CREATE%20and%20DELETE%20to%20avoid%20AddOrUpdate%20method.%20This%20way%20I%20will%20delete%20note%2C%20if%20it%20already%20exists%20for%20collaborator%20and%20then%20call%20create%20method.%20Also%20I%20do%20not%20have%20logic%20for%20deleting%20note%20as%20of%20now%2C%20so%20it%20can%20solve%20both%20moments.%22%2C%20%22created_at%22%3A%20%222016-05-10T12%3A00%3A48Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/14029898%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Rudeka%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ok%2C%20lets%20try%20do%20that%20in%20suggested%20way.%22%2C%20%22created_at%22%3A%20%222016-05-10T12%3A02%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6068068%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/TerraVenil%22%7D%7D%2C%20%7B%22body%22%3A%20%22Fixed%20and%20commited%22%2C%20%22created_at%22%3A%20%222016-05-10T14%3A56%3A11Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/14029898%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Rudeka%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/INoteCommands.cs%3AL6-12%22%7D%2C%20%22Pull%20b34981bfea3cfcb9c7e9f77f6b22c2c54ce4ed1d%20DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/NoteCommands.cs%2015%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r62688415%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22You%20can%20prevent%20using%20this%20verification.%20Can%20you%20explain%20why%3F%22%2C%20%22created_at%22%3A%20%222016-05-10T15%3A01%3A35Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6068068%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/TerraVenil%22%7D%7D%2C%20%7B%22body%22%3A%20%22If%20I%20have%20one%20to%20one%20relation%20between%20Collaborators%20and%20Notes%2C%20then%20it%20can%27t%20be%20more%20than%20one%20note%20for%20collaborator.%20%22%2C%20%22created_at%22%3A%20%222016-05-10T15%3A16%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/14029898%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Rudeka%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20have%20added%20unique%20constraint%20for%20userId%20and%20collaboratorId%2C%20so%20I%20can%20prevent%20using%20SingleOrDefault.%20Fix%20commited%22%2C%20%22created_at%22%3A%20%222016-05-10T17%3A08%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/14029898%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Rudeka%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20know%20but%20anyway.%20If%20user%20was%20deleted%20from%20DB%20that%20concurrency%20exception%20will%20be%20raised%20and%20you%20can%20handled%20it%20in%20another%20way.%20So%20this%20code%20obsolete.%22%2C%20%22created_at%22%3A%20%222016-05-10T17%3A19%3A24Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6068068%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/TerraVenil%22%7D%7D%2C%20%7B%22body%22%3A%20%22Verification%20removed%2C%20new%20commit%20added%22%2C%20%22created_at%22%3A%20%222016-05-11T16%3A12%3A26Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/14029898%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Rudeka%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/NoteCommands.cs%3AL20-42%22%7D%2C%20%22Pull%2058514d3efa0cafa3de3572933475f79bcc9c2513%20DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs%2026%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/ITsvetkoFF/Kv-013/pull/139%23discussion_r63161901%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20don%27t%20undertand%20why%20it%20can%20be%20moved%20to%20extension%20method%3F%20Why%20you%20need%20this%20logic%20inside%20controller%3F%22%2C%20%22created_at%22%3A%20%222016-05-13T10%3A17%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6068068%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/TerraVenil%22%7D%7D%2C%20%7B%22body%22%3A%20%22Yes%2C%20it%20looks%20much%20better%20as%20an%20extension%20method%20after%20previous%20changes%20I%27ve%20made%20according%20to%20your%20comments.%20Moved%20logic%20to%20an%20extension%20method%20and%20commited.%22%2C%20%22created_at%22%3A%20%222016-05-13T10%3A54%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/14029898%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Rudeka%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs%3AL210-238%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'><img src='http://www.codereviewhub.com/site/github-bar.png' height=40></a>
- [ ] <a href='#crh-comment-Pull 56d9f5d8a20945844b5c85d8ad461e6d4470cee9 DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/INoteCommands.cs 5'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/ITsvetkoFF/Kv-013/pull/139#discussion_r62653642'>File: DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/INoteCommands.cs:L6-12</a></b>
- <a href='https://github.com/TerraVenil'><img border=0 src='https://avatars.githubusercontent.com/u/6068068?v=3' height=16 width=16'></a> Explain please why you are decided change method name AddNote -> AddOrUpdateNote? What about SOLID?
- <a href='https://github.com/Rudeka'><img border=0 src='https://avatars.githubusercontent.com/u/14029898?v=3' height=16 width=16'></a> We should have only one note for collaborator from user and I decided not to call separate method for update a note, if one already exists. It is not single responsibility and decided for this simple case we can ignore it.
  But now I think I should better have two methods CREATE and DELETE to avoid AddOrUpdate method. This way I will delete note, if it already exists for collaborator and then call create method. Also I do not have logic for deleting note as of now, so it can solve both moments.
- <a href='https://github.com/TerraVenil'><img border=0 src='https://avatars.githubusercontent.com/u/6068068?v=3' height=16 width=16'></a> Ok, lets try do that in suggested way.
- <a href='https://github.com/Rudeka'><img border=0 src='https://avatars.githubusercontent.com/u/14029898?v=3' height=16 width=16'></a> Fixed and commited
- [ ] <a href='#crh-comment-Pull b34981bfea3cfcb9c7e9f77f6b22c2c54ce4ed1d DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/NoteCommands.cs 15'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/ITsvetkoFF/Kv-013/pull/139#discussion_r62688415'>File: DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/NoteCommands.cs:L20-42</a></b>
- <a href='https://github.com/TerraVenil'><img border=0 src='https://avatars.githubusercontent.com/u/6068068?v=3' height=16 width=16'></a> You can prevent using this verification. Can you explain why?
- <a href='https://github.com/Rudeka'><img border=0 src='https://avatars.githubusercontent.com/u/14029898?v=3' height=16 width=16'></a> If I have one to one relation between Collaborators and Notes, then it can't be more than one note for collaborator.
- <a href='https://github.com/Rudeka'><img border=0 src='https://avatars.githubusercontent.com/u/14029898?v=3' height=16 width=16'></a> I have added unique constraint for userId and collaboratorId, so I can prevent using SingleOrDefault. Fix commited
- <a href='https://github.com/TerraVenil'><img border=0 src='https://avatars.githubusercontent.com/u/6068068?v=3' height=16 width=16'></a> I know but anyway. If user was deleted from DB that concurrency exception will be raised and you can handled it in another way. So this code obsolete.
- <a href='https://github.com/Rudeka'><img border=0 src='https://avatars.githubusercontent.com/u/14029898?v=3' height=16 width=16'></a> Verification removed, new commit added
- [ ] <a href='#crh-comment-Pull b34981bfea3cfcb9c7e9f77f6b22c2c54ce4ed1d DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs 31'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/ITsvetkoFF/Kv-013/pull/139#discussion_r62688613'>File: DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs:L209-235</a></b>
- <a href='https://github.com/TerraVenil'><img border=0 src='https://avatars.githubusercontent.com/u/6068068?v=3' height=16 width=16'></a> Can you do this only by one call to DB?
- <a href='https://github.com/Rudeka'><img border=0 src='https://avatars.githubusercontent.com/u/14029898?v=3' height=16 width=16'></a> Fixed and commited
- [ ] <a href='#crh-comment-Pull d9c874074245403f0fae24e90d3b6e91f41e9c49 DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/NoteCommands.cs 19'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/ITsvetkoFF/Kv-013/pull/139#discussion_r62881279'>File: DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/NoteCommands.cs:L19-36</a></b>
- <a href='https://github.com/TerraVenil'><img border=0 src='https://avatars.githubusercontent.com/u/6068068?v=3' height=16 width=16'></a> So it works? Did you verify it?
- <a href='https://github.com/Rudeka'><img border=0 src='https://avatars.githubusercontent.com/u/14029898?v=3' height=16 width=16'></a> Yes, it works. With EntityState I had to make it tracked first and then call method Remove.
- [ ] <a href='#crh-comment-Pull d9c874074245403f0fae24e90d3b6e91f41e9c49 DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs 29'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/ITsvetkoFF/Kv-013/pull/139#discussion_r62881505'>File: DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs:L209-237</a></b>
- <a href='https://github.com/TerraVenil'><img border=0 src='https://avatars.githubusercontent.com/u/6068068?v=3' height=16 width=16'></a> To many works inside one method....Please refactor it.
- <a href='https://github.com/Rudeka'><img border=0 src='https://avatars.githubusercontent.com/u/14029898?v=3' height=16 width=16'></a> Moved query to context to action method and passed it as an argument. Commit added
- [ ] <a href='#crh-comment-Pull 58514d3efa0cafa3de3572933475f79bcc9c2513 DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs 26'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/ITsvetkoFF/Kv-013/pull/139#discussion_r63161901'>File: DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs:L210-238</a></b>
- <a href='https://github.com/TerraVenil'><img border=0 src='https://avatars.githubusercontent.com/u/6068068?v=3' height=16 width=16'></a> I don't undertand why it can be moved to extension method? Why you need this logic inside controller?
- <a href='https://github.com/Rudeka'><img border=0 src='https://avatars.githubusercontent.com/u/14029898?v=3' height=16 width=16'></a> Yes, it looks much better as an extension method after previous changes I've made according to your comments. Moved logic to an extension method and commited.

<a href='https://www.codereviewhub.com/ITsvetkoFF/Kv-013/pull/139?mark_as_completed=1'><img src='http://www.codereviewhub.com/site/github-mark-as-completed.png' height=26></a>&nbsp;<a href='https://www.codereviewhub.com/ITsvetkoFF/Kv-013/pull/139?approve=1'><img src='http://www.codereviewhub.com/site/github-approve.png' height=26></a>&nbsp;<a href='https://github.com/ITsvetkoFF/Kv-013/pull/139'><img src='http://www.codereviewhub.com/site/github-refresh.png' height=26></a>
<a href='#crh-end'></a>

Rudeka committed:
Add userId to collaborators

Rudeka committed:
Add note update if already exists

	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.Notes.Tests/TestsForControllers/NoteControllerCreateNoteTests.cs
	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/INoteCommands.cs
	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/NoteCommands.cs
	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Controllers/NoteController.cs

	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.Notes.Tests/TestsForControllers/NoteControllerCreateNoteTests.cs
	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/INoteCommands.cs
	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/NoteCommands.cs
	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Controllers/NoteController.cs

ITsvetkoFF commented:
This pull request is now managed with CodeReviewHub. <a href='https://github.com/ITsvetkoFF/Kv-013/pull/139'>Refresh</a>.<a href='#crh-update-none'></a>

TerraVenil commented on DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/INoteCommands.cs:
```diff
@@ -6,6 +6,6 @@ namespace GitHubExtension.Notes.WebApi.Commands
     public interface INoteCommands
     {
-        Task AddNote(Note noteEntity);
+        Task AddOrUpdateNote(Note noteEntity);
```
Explain please why you are decided change method name AddNote -> AddOrUpdateNote? What about SOLID?

Rudeka replied:
We should have only one note for collaborator from user and I decided not to call separate method for update a note, if one already exists. It is not single responsibility and decided for this simple case we can ignore it.
But now I think I should better have two methods CREATE and DELETE to avoid AddOrUpdate method. This way I will delete note, if it already exists for collaborator and then call create method. Also I do not have logic for deleting note as of now, so it can solve both moments.

TerraVenil replied:
Ok, lets try do that in suggested way.

Rudeka replied:
Fixed and commited

Rudeka committed:
Separated AddOrUpdate method to Add and Delete methods

	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.Infrastructure.Constants/RouteConstants.cs
	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.Notes.Tests/TestsForControllers/NoteControllerCreateNoteTests.cs
	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.Notes.Tests/TestsForRoutes/TestRoutesMappers/ForGetNoteForCollaborator.cs
	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/INoteCommands.cs
	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/NoteCommands.cs
	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Controllers/NoteController.cs
	modified:   DevelopGitHub/GitHubExtension/GitHubExtension.sln

TerraVenil commented on DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/NoteCommands.cs:
```diff
@@ -18,5 +20,22 @@ public NoteCommands(NoteContext notesContext)
+
+        public async Task<bool> DeleteNote(string userId, string collaboratorId)
+        {
+            var note = _notesContext.Notes.SingleOrDefault(n => n.UserId == userId 
```
You can prevent using this verification. Can you explain why?

Rudeka replied:
If I have one to one relation between Collaborators and Notes, then it can't be more than one note for collaborator.

Rudeka replied:
I have added unique constraint for userId and collaboratorId, so I can prevent using SingleOrDefault. Fix commited

TerraVenil replied:
I know but anyway. If user was deleted from DB that concurrency exception will be raised and you can handled it in another way. So this code obsolete. Clarify?

Rudeka replied:
Verification removed, new commit added

TerraVenil commented on DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs:
```diff
@@ -207,5 +209,26 @@ private void AddClaimsToCookie(Claim[] claims)
+
+            foreach (var collaborator in gitHubCollaboratorsExceptUser)
+            {
+                var user = _userManager.FindByGitHubId(collaborator.Id);
```
Can you do this only by one call to DB?

Rudeka replied:
Fixed and commited

Rudeka committed:
Change .SingleOrDefault to .FirstOrDefault

Rudeka committed:
Refactor single query to db to add userId to collaborator

Rudeka committed:
Add separate routes for GET and DELETE

Rudeka committed:
Remove verification before delete

TerraVenil commented on DevelopGitHub/GitHubExtension/GitHubExtension.Notes.WebApi/Commands/NoteCommands.cs:
```diff
@@ -18,5 +19,17 @@ public NoteCommands(NoteContext notesContext)
+                Id = noteId
+            };
+
+            _notesContext.Entry(note).State = EntityState.Deleted;
```
So it works? Did you verify it?

Rudeka replied:
Yes, it works. With EntityState I had to make it tracked first and then call method Remove.

TerraVenil commented on DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs:
```diff
@@ -207,5 +209,28 @@ private void AddClaimsToCookie(Claim[] claims)
+        {
+            var collaborators = new List<CollaboratorWithUserIdModel>();
+
+            var users = _securityContextQuery.GetAllUsers();
```
To many works inside one method....Please refactor it.

Rudeka replied:
Moved query to context to action method and passed it as an argument. Commit added

Rudeka committed:
Moved query to context to action method and passed it as a parameter

TerraVenil commented on DevelopGitHub/GitHubExtension/GitHubExtension.Security.WebApi/Controllers/RepositoryController.cs:
```diff
@@ -207,5 +210,28 @@ private void AddClaimsToCookie(Claim[] claims)
             }
         }
+
+        IEnumerable<CollaboratorWithUserIdModel> AddUserIdToCollaboratorIfExists(
```
I don't undertand why it can be moved to extension method? Why you need this logic inside controller?

Rudeka replied:
Yes, it looks much better as an extension method after previous changes I've made according to your comments. Moved logic to an extension method and commited.

Rudeka committed:
Moved adding userId to collaborator to an extension method
